graphql-server/src/utils/cache.ts

```diff
@@ -0,0 +1,218 @@
+  code: string;
+  color: string;
+  externalReferences: Record<string, unknown>;
+  history: string[];
+  level: number;
+  mainType: string;
+  name: string;
+  parent: string;
+  precursors: string[];
+  revocations: string[];
+  tissue: string;
+};
+
+export async function initializeInMemoryCache() {
+  const inMemoryCache = new NodeCache();
+
+  // Warm up the cache
+  await updateOncotreeCache(inMemoryCache);
+  await updateSamplesCache(inMemoryCache);
+  logCacheStats(inMemoryCache);
+
+  // Add cache item expiration handlers
+  // (node-cache checks for expired items and runs this event listener every 10m by default)
+  inMemoryCache.on("expired", async (key) => {
+    if (key === ONCOTREE_CACHE_KEY) {
+      await updateOncotreeCache(inMemoryCache);
+      logCacheStats(inMemoryCache);
+    }
+    if (key === SAMPLES_CACHE_KEY) {
+      await updateSamplesCache(inMemoryCache);
+      logCacheStats(inMemoryCache);
+    }
+  });
+
+  return inMemoryCache;
+}
+
+export async function updateOncotreeCache(inMemoryCache: NodeCache) {
+  const oncotreeApiData = await fetchOncotreeApiData();
+  if (!oncotreeApiData) return;
+
+  const oncotreeCache = oncotreeApiData.reduce((acc, tumor) => {
+    acc[tumor.code] = { name: tumor.name, mainType: tumor.mainType };
+    return acc;
+  }, {} as OncotreeCache);
+
+  // Add to cache the oncotree codes found in Neo4j but not in Oncotree API's response
+  // (These codes are likely old codes that have been renamed in the Oncotree database)
+  const oncotreeCodesInNeo4j = await getOncotreeCodesFromNeo4j();
+  oncotreeCodesInNeo4j?.forEach((code) => {
+    if (!oncotreeCache.hasOwnProperty(code)) {
+      oncotreeCache[code] = { name: "N/A", mainType: "N/A" };
+    }
+  });
+
+  inMemoryCache.set(ONCOTREE_CACHE_KEY, oncotreeCache, ONCOTREE_CACHE_TTL);
+}
+
+async function fetchOncotreeApiData() {
+  try {
+    const response = await fetch(props.oncotree_api, {
+      headers: {
+        Accept: "application/json",
+      },
+    });
+    if (!response.ok) {
+      throw new Error(
+        `Failed to fetch the Oncotree API: ${response.statusText}`
+      );
+    }
+    return (await response.json()) as Promise<OncotreeApiTumorType[]>;
+  } catch (error) {
+    if (error instanceof Error) {
+      console.error(`Error: ${error.message}`);
+    } else {
+      console.error("Unknown error occurred while fetching Oncotree data");
+    }
+    return null;
+  }
+}
+
+async function getOncotreeCodesFromNeo4j() {
+  const session = neo4jDriver.session();
+  try {
+    const result = await session.writeTransaction((tx) =>
+      tx.run(`
+        MATCH (sm:SampleMetadata) RETURN DISTINCT sm.oncotreeCode AS oncotreeCode
+      `)
+    );
+    return result.records
+      .map((record) => record.get("oncotreeCode"))
+      .filter(Boolean) as string[];
+  } catch (error) {
+    if (error instanceof Error) {
+      console.error(error.message);
+    }
+  } finally {
+    await session.close();
+  }
+}
```

Comment:
Some simplification of code but most of this remains the same as it was in `oncotree.ts`.